[PlutusTx] Optimize the 'Eq' instance of 'Value'

effectfully opened the pull request:
Do not look here, currently this PR contains a lot of irrelevant noise, which will be removed once #5566 is in.

effectfully commented on plutus-tx/src/PlutusTx/SortedMap.hs:
```diff
@@ -0,0 +1,78 @@
+        case k0 `compare` k of
+            LT -> (k0, v0) : kvs
+            -- TODO: make this @traceError duplicateElements@.
+            EQ -> (k, v0) : kvs'
```
I'm not making the change in this PR, so that the budgets comparison is fair. I'll make the change and I'll update the golden tests in general, since the current ones were written with the assumption that duplicate currencies/tokens are allowed.

effectfully commented:
This fixes the performance part of #5135, but we need tests for this stuff too, it's planned for a future PR.

bezirg commented on plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:
```diff
@@ -368,6 +364,118 @@ split (Value mp) = (negate (Value neg), Value pos) where
+    | MatchFailure
+
+{-# INLINE matchKVs #-}
+{- | Take a function checking whether a value is zero\/empty, a function checking /structural/
```
```suggestion
{- | Takes
```

effectfully replied:
I like the first sentence to have imperative mood. I don't mind changing it, but I do that for virtually every Haddock that I wrote, so if I do it only here, it would only make things inconsistent.

I don't mind changing this habit, if it's something that we want. @zliu41 @michaelpj do you have an opinion?

bezirg commented on plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:
```diff
@@ -368,6 +364,118 @@ split (Value mp) = (negate (Value neg), Value pos) where
+
+{-# INLINE matchKVs #-}
+{- | Take a function checking whether a value is zero\/empty, a function checking /structural/
+equality of two values and two key-value lists and return the result of matching the lists
```
```suggestion
returns...
```

bezirg commented on plutus-tx/src/PlutusTx/SortedMap.hs:
```diff
@@ -0,0 +1,79 @@
+
+{-# INLINE singleton #-}
+singleton :: k -> v -> SortedMap k [v]
+singleton ~k ~v = UnsafeSortedMap [(k, [v])]
```
Why the arguments to this function are lazy? Also aren't we supposed to use Strict nowadays to the whole of plutustx prelude?

bezirg commented on plutus-tx/src/PlutusTx/SortedMap.hs:
```diff
@@ -0,0 +1,79 @@
+    go ((k1, v1) : kvs1) ((k2, v2) : kvs2) =
+        -- As with 'matchKVs' we check equality of all the keys first and only then check equality
+        -- of values.
+        if k1 == k2
+            then if go kvs1 kvs2
+                then eqV v1 v2
+                else False
+            else False
```
```suggestion
        if k1 == k2 && go kvs1 kvs2
        then eqV v1 v2
        else False
```

michaelpj replied:
This will perform worse, remember that `&&` is not lazy

bezirg replied:
Omg i forgot that (&&) in plutustx is not short-circuitting

michaelpj replied:
application is still strict. This is a weird corner and I think Ziyang wants to make it somehow work, but at the moment `f x y` evaluates x and y always.

effectfully replied:
> Oh wait, it is short-circuitting , I don't know then what you mean

Only if GHC inlines it, which we probably can't 100% rely upon.

bezirg commented on plutus-tx/src/PlutusTx/SortedMap.hs:
```diff
@@ -0,0 +1,79 @@
+    :: forall k v. Eq k
+    => (v -> Bool) -> (v -> v -> Bool) -> SortedMap k v -> SortedMap k v -> Bool
+eqWith ~is0 ~eqV = coerce go where
+    go :: [(k, v)] -> [(k, v)] -> Bool
+    go [] []   = True
+    go [] kvs2 = all (is0 . snd) kvs2  -- If one of the lists is empty then all elements in the
+    go kvs1 [] = all (is0 . snd) kvs1  -- other list need to be zero for lists to be equal.
+    go ((k1, v1) : kvs1) ((k2, v2) : kvs2) =
+        -- As with 'matchKVs' we check equality of all the keys first and only then check equality
```
This looks like it shares some logic with matchKVs , but I don't have any suggestion on how to share it. Actually maybe it is better to keep them separate

bezirg commented on plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:
```diff
@@ -368,6 +364,118 @@ split (Value mp) = (negate (Value neg), Value pos) where
+                then
+                    -- Check equality of values that come from the common key prefix of the original
+                    -- lists.
+                    all (uncurry (==)) valPairs
```
Here valpairs are maps themselves. Don't you have to do proper sorted equality check (i.e. eqMapVia)?

effectfully replied:
Thank you for spotting the bug, it's fixed now.

michaelpj commented on plutus-tx/src/PlutusTx/SortedMap.hs:
```diff
@@ -0,0 +1,79 @@
+-- The pragma trades a bit size for potentially plenty of budget.
+{-# INLINE eqWith #-}
+-- | Check equality of 'SortedMap's by matching the underlying lists pointwise.
+eqWith
```
This is a bit `Value`-specific: the `is0` function only really makes sense if you are working with a map that has the structure that `Value` does, as opposed to a generic sorted map. That's not necessarily a problem but we should say so.

effectfully replied:
Moved it to `Value.hs` and got rid of `SortedMap` entirely.

michaelpj reviewed (approved):
Given how many intermediate structures this must create, I'm amazed it performs okay, but I guess that just means the old one was truly appalling :grimacing:

effectfully commented on plutus-tx/src/PlutusTx/SortedMap.hs:
```diff
@@ -0,0 +1,79 @@
+    go []             = UnsafeSortedMap []
+    go ((k, v) : kvs) = unsafeInsertOneUnique k v $ go kvs
+
+-- The pragma trades a bit size for potentially plenty of budget.
```
![Screenshot from 2023-10-10 18-07-07](https://github.com/input-output-hk/plutus/assets/10480926/f332874f-761e-41a9-80ed-07cddc6b353e)

> Maybe GHC is doing something?

Quite likely.

effectfully replied:
... and

![Screenshot from 2023-10-10 18-14-09](https://github.com/input-output-hk/plutus/assets/10480926/9bd25adb-8198-4056-bb6e-667fd974ca99)

kwxm commented on plutus-tx/src/PlutusTx/SortedMap.hs:
```diff
@@ -0,0 +1,79 @@
+    go [] []   = True
+    go [] kvs2 = all (is0 . snd) kvs2  -- If one of the lists is empty then all elements in the
+    go kvs1 [] = all (is0 . snd) kvs1  -- other list need to be zero for lists to be equal.
+    go ((k1, v1) : kvs1) ((k2, v2) : kvs2) =
```
Later: I think this is actually a bug.  See the example in Value.hs.

kwxm commented on plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:
```diff
@@ -368,6 +364,118 @@ split (Value mp) = (negate (Value neg), Value pos) where
+        -- Otherwise the keys in the lists have diverged and we return the remaining parts.
+        | otherwise = MatchPartial [] kvs1 kvs2
+
+{-# INLINEABLE eq #-}
```
I think this doesn't actually work: there's a problem with non-trailing entries with zero values.  In the code below, `v1 == v2` is `False` but `v2 == v3` is `True`.

```
{-# LANGUAGE OverloadedStrings #-}

import PlutusLedgerApi.V1.Value
import PlutusTx.AssocMap as Map
    
ada :: CurrencySymbol
ada = ""

a :: TokenName
a = "aa"

b :: TokenName
b = "bb"

c :: TokenName
c = "cc"
    
v1 :: Value
v1 = Value $ Map.fromList [(ada, Map.fromList [(a,0),(b,100)])]

v2 :: Value
v2 = Value $ Map.fromList [(ada, Map.fromList [(b,100)])]

v3 :: Value
v3 = Value $ Map.fromList [(ada, Map.fromList [(b,100),(c,0)])]
```

effectfully replied:
Yup, thanks a lot for debugging the issue and providing the example. It's a bug indeed: I forgot to replicate the `is0` logic of `matchKVs` in `eqKVs`. My imaginary tests would catch the bug, but I'm rushing (for PI planning reasons) and imaginary tests don't get run in the real world, unfortunately. I've fixed the bug now and I'll add proper tests in a subsequent PR.

kwxm reviewed (commented):
I think there's still a bug in this.  I found it quite hard to understand what the code was supposed to be doing and I wonder if we could try to convert it to Agda to get some more assurance of correctness.  This would be a good example to try because it's pretty much standalone code where we don't have to rely on lots of external stuff.  @mjaskelioff do you have an opinion on this?

effectfully commented on plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:
```diff
@@ -368,6 +364,118 @@ split (Value mp) = (negate (Value neg), Value pos) where
+        -- If keys are equal
+        | k1 == k2 =
+            -- then continue checking equality of spines.
+            case go kvs1' kvs2' of
```
It does a bit, as per #5530, not having tail recursion is better than having an extra argument, for example. So even if it costs anything, it's pretty negligible still.

effectfully commented on plutus-tx/src/PlutusTx/SortedMap.hs:
```diff
@@ -0,0 +1,79 @@
+    go [] kvs2 = all (is0 . snd) kvs2  -- If one of the lists is empty then all elements in the
+    go kvs1 [] = all (is0 . snd) kvs1  -- other list need to be zero for lists to be equal.
+    go ((k1, v1) : kvs1) ((k2, v2) : kvs2) =
+        -- As with 'matchKVs' we check equality of all the keys first and only then check equality
```
Now in the same file.

effectfully commented on plutus-tx/src/PlutusTx/SortedMap.hs:
```diff
@@ -0,0 +1,79 @@
+    go []             = UnsafeSortedMap []
+    go ((k, v) : kvs) = unsafeInsertOneUnique k v $ go kvs
+
+-- The pragma trades a bit size for potentially plenty of budget.
```
*a bit of size. Fixed.

effectfully commented on plutus-tx/src/PlutusTx/SortedMap.hs:
```diff
@@ -0,0 +1,79 @@
+{-# INLINE unsafeFromListUnique #-}
+-- | Turn a list into a 'SortedMap' assuming all of its keys are unique (if they are not, the
+-- function throws).
+unsafeFromListUnique :: forall k v. Ord k => [(k, v)] -> SortedMap k v
```
> Did you try any other sorting method here?

Yes, I've tried the one we have in `PlutusTx.List` and it basically doesn't finish. It's order of magnitudes worse, I've tested it on much smaller inputs and it still was much worse. I've no idea why, I investigated it shortly, but couldn't quickly find the culprit, so I moved on. I'll make a ticket.

effectfully replied:
> Also, would it be worth trying to be tail recursive here and/or in the insertion function?

According to #5530 probably not.

effectfully commented on plutus-ledger-api/test-plugin/Spec/golden/9.2/Long.stat.golden:
> Do you know how big the compiled function is, eg as a fraction of the maximum transaction size (16384 bytes)?

I don't know, but here's the diff of [plutus-benchmark/marlowe/test/semantics/semantics.size.golden](https://github.com/input-output-hk/plutus/pull/5577/files#diff-8d29f73c346b97d05a7c8a7940d9cbdb90a59b2e52763eec7dfdcc82edb6fadd) in this PR:

```diff
@@ -1 +1 @@
-13603
+14911
```

Does not look very good... Perhaps I should optimize for size more than for CPU and memory. I'll look into it in a future PR.

effectfully replied:
I've managed to make it

```diff
-14911
+14617
```

It's what I get if I remove all `INLINE` pragmas. So I need to either define `eqKVs` (used to be `eqWith`) in terms of `matchKVs` to share more code or change the algorithm, if I want to bring that number further down. I have an idea on how to improve the algorithm generally, so I'll do that in a subsequent PR (after adding tests) instead of dumbing the current algorithm down in this PR.

effectfully replied:
Ah, the cost for that size reduction is

```diff
index f5a2f50f0..93fa88acb 100644
--- a/plutus-benchmark/script-contexts/test/checkScriptContextEqualityTerm-20.eval.golden
+++ b/plutus-benchmark/script-contexts/test/checkScriptContextEqualityTerm-20.eval.golden
@@ -1,2 +1,2 @@
-({cpu: 243201611
-| mem: 951066})
\ No newline at end of file
+({cpu: 268570611
+| mem: 1061366})
```

So not very good, but not too bad either. Size is probably more important.

effectfully commented on plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:
```diff
@@ -368,6 +364,118 @@ split (Value mp) = (negate (Value neg), Value pos) where
+        -> Map.Map TokenName Integer
+        -> Bool
+    eqMapVia asSortedMap (Map.toList -> tokens1) (Map.toList -> tokens2) =
+        SortedMap.eqWith (== 0) (==) (asSortedMap tokens1) (asSortedMap tokens2)
```
I've removed the `INLINE` pragma from that function, it did help a bit, see the thread above. Thanks for bringing it up!

effectfully commented on plutus-tx/src/PlutusTx/SortedMap.hs:
```diff
@@ -0,0 +1,79 @@
+-- | Turn a list into a 'SortedMap' assuming all of its keys are unique (if they are not, the
+-- function throws).
+unsafeFromListUnique :: forall k v. Ord k => [(k, v)] -> SortedMap k v
+unsafeFromListUnique = go where
```
Probably not. We still have to keep this logic for `matchKVs` and replicating it for `eqKVs` is straightforward, while making a perfectly generic insertion sort aware of zeroes is kinda nasty.

effectfully commented:
I addressed most of the comments. I'm going to hit "merge", so that we can claim this is done and I'll start working on proper property tests for it, then I'll revise the unit tests that we have and that were written with a different set of constraints in mind, then I'll try tweaking the main algorithm to make it smaller and possibly more efficient.

michaelpj commented:
I think maybe we shouldn't have merged this. It's tricky code and we found some bugs, so we probably want to make sure we get the tests in. But we're doing releases every sprint now! So it's quite possible that this could get into a release in a buggy state. In fact, I think @kwxm is planning to do a release today? So I think we should make sure this isn't in the release, or delay the release until the tests land.

kwxm commented:
> I think @kwxm is planning to do a release today?

Probably tomorrow (Thursday).

effectfully commented:
I've reverted the merge. @kwxm feel free to create a new release, everything should be fine. I'll reopen the PR or open another one once I have tests.
